tasks.c

```diff
@@ -3808,6 +3808,8 @@ void vTaskSuspendAll( void )
 
         /* This must only be called from within a task. */
         portASSERT_IF_IN_ISR();
+        /* This must never be called from inside a critical section */
+        configASSERT( portGET_CRITICAL_NESTING_COUNT() == 0 );
```

Comment:
@aggarg @RichardBarry

Is this change alright? I know some ports such as the [ARM_CRx_No_GIC](https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V11.0.1/portable/GCC/ARM_CRx_No_GIC/portASM.S#L60) push the critical nesting count as part of the task context:
```
    /* Push the critical nesting count. */
    LDR     R2, ulCriticalNestingConst
    LDR     R1, [R2]
    PUSH    {R1}
```
I'm worried that this change might be a breaking change for some ports/applications and was hoping for a deeper look into this PR from you two

Reply:
portGET_CRITICAL_NESTING_COUNT() is okay to me. Sharing my observation here.
Critical nesting count should only be modified in critical section and context switch inside a critical section is not allowed in SMP. Use the following scenarios for discussion:
1. portGET_CRITICAL_NESTING_COUNT() is greater than 1. This task is in a critical section and can read or update the critical nesting count without problem.
2. portGET_CRITICAL_NESTING_COUNT() is 0. The task A may get preempted by task B while reading critical nesting count. Critical nesting count will be restore to 0 when the task B left critical section.